### DISCUSSION TITLE
How can I test a route in Remix? Should I test it?

### DISCUSSION THREAD

**Discussion by willpinha** [remix-run/remix#9935] [Q&A]
I'm creating tests for my application using @remix-run/testing, Playwright, MSW and React Testing Library. So far, the types of tests I've been able to perform are:

- Test an action/loader individually, as pointed out [in this article](https://sergiodxa.com/tutorials/test-remix-loaders-and-actions), along with mocks for API requests
- Create a stub with @remix-run/testing to test the behavior of a page with mocked loader/action

My question is, so far I've been able to test loader/action or a page individually, but I still don't know how to test both together. Is there any way to test an entire route, including the rendering component and the loader/action?

One way I thought of is to test the loader/action and the route component individually, but my fear of testing loader/action individually is that I would be testing implementation, not behavior, as explained [in this article](https://kentcdodds.com/blog/testing-implementation-details)

Do my concerns make sense? How do you normally test your Remix applications?

**Comment by admirsaheta** [comment]
Your concerns about testing loaders/actions and route components individually are valid. Testing them in isolation can sometimes lead to focusing on implementation details rather than the overall behavior. However, you can approach testing in Remix in a way that balances the need to verify behavior while also ensuring that your loaders/actions and components work as expected.

To test both the route (rendering component) and the loader/action together, you can take an integration testing approach. Here’s how you can achieve this:

```ts
import { render } from "@testing-library/react";
import { setupServer } from "msw/node";
import { handlers } from "../mocks/handlers"; // Your MSW handlers
import { MemoryRouter, Route } from "@remix-run/react";
import MyComponent from "../routes/my-route";

const server = setupServer(...handlers);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("renders the route with loader data", async () => {
  const { getByText } = render(
    <MemoryRouter initialEntries={["/my-route"]}>
      <Route path="/my-route" element={<MyComponent />} />
    </MemoryRouter>
  );

  // Assuming your loader fetches some data and renders it
  expect(getByText("Expected Loader Data")).toBeInTheDocument();
});
```
For more complex interactions, end-to-end testing with Playwright allows you to test the entire route, including the loader/action, in a real browser environment.

```ts
import { test, expect } from '@playwright/test';

test('full route test', async ({ page }) => {
  await page.goto('/my-route');

  // Verify loader data is rendered
  await expect(page.locator('text=Expected Loader Data')).toBeVisible();

  // Simulate user interaction that triggers the action
  await page.click('button#do-something');

  // Verify the result of the action
  await expect(page.locator('text=Expected Action Result')).toBeVisible();
});
```

**Reply by roger-tbg** [reply]
This answer is a month old -- but I don't think MemoryRouter is inside of @remix-run/react.

**Reply by kiliman** [reply]
I created a document that shows how to do unit tests, route tests, and end-to-end testing with Remix.

https://github.com/kiliman/remix-vite-template/blob/main/docs/testing.md

**Reply by roger-tbg** [reply]
Thank you!  i'm having possibly a related issue with trying to do the above.. if I ever try to 'stub out the loader' then the route component never renders:
https://github.com/roger-tbg/my-remix-vite-app

(just renders:
``` <body><div/></body>```
)

**Reply by kiliman** [reply]
Hmm.. not sure. Looks like `screen.debug()` isn't quite right. I tried it with my counter example and it also shows `<body><div/></body>`, but when I console.log(button), it outputs a valid React element.

<img width="738" alt="image" src="https://github.com/user-attachments/assets/0d56ec91-4ab5-42ae-a230-61ddcde2e79c">

Also, if I use a fake loader, it appears to be rendering it correctly.

```ts
async function fakeLoader() {
  return json({ count: 123 });
}

test("counter increments when clicked", async () => {
  const App = createRemixStub([
    {
      path: "/counter",
      Component: Counter,
      loader: fakeLoader,
      action,
    },
  ]);
  await render(<App initialEntries={["/counter"]} />);
  console.log(screen.debug());
  const button = await screen.findByRole("button", { name: /count:/i });
  console.log(button);
});
```
<img width="622" alt="image" src="https://github.com/user-attachments/assets/a883d3b2-c104-4c3b-9585-b1928b5ce0bd">
